### ISSUE TITLE
watchexec takes 8s to start, does 2M syscalls across 148 threads

### ISSUE THREAD

**Issue by kentonv** [watchexec/watchexec#377]
I'm using watchexec 1.20.5 on Linux. It works great! Except for one thing...

I have a directory structure like this:

```
src/   # my code
deps/  # literally like 200,000 files
```

When I run `watchexec` I do it like this:

```
watchexec --watch src make
```

`watchexec` spends about 8 seconds scanning through the entirety of `deps`. The `-vvvv` log, which is 528 megabytes long, suggests it is looking for "ignore files". The various `--no-*-ignore` options don't seem to change anything.

During this time, according to `strace`, `watchexec` apparently creates 148 threads which perform 2,458,067 system calls, of which 2,337,801 are futex operations. It also stat()s 44,270 files and opens 14,634 of them. After all that, it sets up 26 inotify watches.

Could `watchexec` be made to scan only the directory I asked it to watch?

**Comment by passcod** [comment]
That's [Known Issue Number 2](https://github.com/watchexec/watchexec/discussions/332)! You can work around for now with `--project-origin src`, or by starting watchexec in `src` and setting the cwd of the command with `--workdir ..`.

**Comment by kentonv** [comment]
Oops, sorry I missed that! Thanks for the tip, `--project-origin` does seem to solve the problem.

The excessive context switching via futex operations might be worth looking into as an optimization opportunity -- it seems to me that even without behavioral changes, some time could probably be saved there. But feel free to close this if you prefer.

**Comment by passcod** [comment]
That's true, yes, we could likely get much better performance by optimising where file I/O happens. I'll have a specific look at this.

**Comment by AaronFriel** [comment]
@passcod is .gitignore parsing currently a depth-first search, when it should be breadth-first starting at the root for project origin detection, and then recursively applying ignore files in descent?

Ignore files only apply to their subtrees, yet after installing an NPM package, I see traces like the following:

```
  2022-09-05T20:14:42.704608Z TRACE watchexec::ignore::files: found a file, path: "/home/friel/c/github.com/pulumi/pulumi/.gitignore"
  2022-09-05T20:14:42.704642Z TRACE watchexec::ignore::filter: reading ignore file, file: IgnoreFile { path: "/home/friel/c/github.com/pulumi/pulumi/.gitignore", applies_in: Some("/home/friel/c/github.com/pulumi/pulumi"), applies_to: Some(Git) }
  2022-09-05T20:14:42.704761Z TRACE watchexec::ignore::filter: adding ignore line, line: "**/node_modules/"
...
    in watchexec::ignore::files::visit_path with path: "/home/friel/c/github.com/pulumi/pulumi/sdk/nodejs/node_modules/tsconfig-paths/src/__tests__"
    in watchexec::ignore::filter::check_dir with path: "/home/friel/c/github.com/pulumi/pulumi/sdk/nodejs/node_modules/tsconfig-paths/src/__tests__/data"
    in watchexec::ignore::files::dir_entry with path: "/home/friel/c/github.com/pulumi/pulumi/sdk/nodejs/node_modules/tsconfig-paths/src/__tests__/data"
    in watchexec::ignore::files::visit_path with path: "/home/friel/c/github.com/pulumi/pulumi/sdk/nodejs/node_modules/tsconfig-paths/src/__tests__"
```

If ignore files follow the `man gitignore` spec, then there's no point in crawling any folder that matches a previously-discovered filter.

This seems like it would be possible for me to make a contribution here. The product I work on, Pulumi, recently shipped a a change to embed `watchexec` to support `pulumi watch` functionality. See: https://github.com/pulumi/watchutil-rs

**Comment by passcod** [comment]
Yes, that's absolutely the case. I'd love a contribution! The work would be in the ignore-files crate and filterer, which get used in the relevant places in watchexec.

**Comment by matu3ba** [comment]
I have a similar issue with `fd -tf . 'src/' | wc -l` returning 75.
However, jabbascript (angular in my case) has a giantic filesize.

I did use `watchexec -w src/ 'ng build && stuff_to_install'`.

**Comment by Shados** [comment]
Aside from any other possible optimisations, shouldn't it be fairly trivial to make watchexec skip looking for ignore files if all of `--no-default-ignore --no-global-ignore --no-project-ignore --no-vcs-ignore` are set?

**Comment by passcod** [comment]
Yes. That would be a conditional around here, if you want to PR this: https://github.com/watchexec/watchexec/blob/main/crates/cli/src/filterer/globset.rs#L23

**Comment by pesterhazy** [comment]
I also ran into this (very significant) problem

`watchexec` takes 8s to start, but when using `--project-origin $PWD`, it's down to 1s

**Comment by infogulch** [comment]
It seems "scanning and watching the entire filesystem for files that could slightly modify the behavior of the program" is more anti-feature than feature for a small utility like this.

**Comment by pesterhazy** [comment]
While a more thorough fix would be great, `--no-discover-ignore` works well for us in the meantime as a workaround

**Comment by t3hmrman** [comment]
Hey @kentonv a fix for this issue *should* have made it into [`watchexec` v1.25.0+](https://github.com/watchexec/watchexec/releases/tag/v1.25.0), would you mind trying the newer version and see if your problem is fixed?

**Comment by passcod** [comment]
I've measured a bit myself and it's not really fixed, though your improvements did help.

However, one larger issue than the ignores, which I hadn't twigged on until I started measuring in earnest, is the notify watches, which are recursed by notify and not by us.

For example looking at watchexec with `RUST_LOG=notify=trace` shows e.g. notify adding watches for the entire .git folder recursively, among other things; fixing that involves writing the watch recursion into watchexec and using non-recursive watches at the notify boundary; I've written up surface notes on the complexities therein here: https://cohost.org/watchexec/post/4070897-early-feature-concep

It turns out that this particular issue here is an onion of causes adding up to a large negative effect, rather than a single buggy layer to be fixed!

**Comment by simonmichael** [comment]
For the record, this confused me greatly for a while at https://github.com/watchexec/watchexec/issues/864#issuecomment-2889241150. More details there.

- While testing something unrelated, I naively ran watchexec in $HOME, no avoid interference from a project's .gitignore.
- When things didn't work as expected, I made sure there was no .gitignore, .ignore, _darcs, .hgrc etc etc in $HOME or any parent directory.
- I did not expect that watchexec was also searching for .git in every subdirectory!
- I did not realise that it was taking a long time to search for gitignores, because there's nothing indicating that.

In my case it was searching for about 40s before processing events (I feel that it varied, I'm pretty sure another time it took closer to 2m.) My home dir currently holds about half a terabyte, and watchexec was finding about 2500 .gitignore files, including 1500 from cached crates in .cargo, and a few hundred each in .go and in .pub-cache (dart).

One possible UX workaround would be to print a warning if this phase is taking more than a few seconds.

**Comment by passcod** [comment]
FTR I've estimated this at about 20 hours of effort on my worklist: https://github.com/watchexec/watchexec/discussions/906

**Comment by passcod** [comment]
As of 2.7.0 I'm calling this fixed. It didn't land in one single change, of course.

---

In my code directory, which has a lot of stuff, most of which is gitignored and should never be traversed, `-1` measures startup time:

```console
$ cd ~/code
$ tree | tail -n1
152380 directories, 2023278 files
$ time watchexec -1 echo
$ strace -fc watchexec -1 echo
```

- 1.20.5 (2022): 42 secs, 7M syscalls, 3M stats, 700k opens
- 2.0.0 (2024): 17 secs, 11M syscalls, 3M stats, 400k opens
- 2.7.0 (2026): 2 secs, 1M syscalls, 110k stats, 38k opens
